Lib/Search/ElasticSearch/Client.php

```diff
@@ -162,9 +163,32 @@ public function removeDocuments(ClassMetadata $class, array $documents)
      */
     public function removeAll(ClassMetadata $class, $query = null)
     {
-        $type = $this->getIndex($class->getIndexForRead())->getType($class->type);
+        $index = $class->getIndexForRead();
         $query = $query ?: new MatchAll();
-        $type->deleteByQuery($query);
+        $this->deleteByScanScroll($class, $query, $index);
+    }
+
+    /**
+     * Perform a scan and scroll through the results, bulk deleting by result ids
+     *
+     * @param ClassMetadata $class
+     * @param AbstractQuery $scanQuery
+     * @param string        $index
+     */
+    private function deleteByScanScroll(ClassMetadata $class, AbstractQuery $scanQuery, $index) {
+        $type = $this->getIndex($index)->getType($class->type);
+        $query = Query::create($scanQuery);
+        $results = $this->scan($query, [$class]);
+
+        foreach($results as $collection) {
+            $resultIds = array();
+            foreach($collection->getResults() as $result) {
+                $resultIds[] = $result->getId();
+            }
+            $this->client->deleteIds($resultIds, $index, $type);
+        }
+        // Refresh index to clear out deleted ID information
+        $this->getIndex($index)->refresh();
```

Comment:
Right. Adding a refresh option to `Client::removeDocuments()` and `Client::removeAll()` and only refreshing at the end of `Client::deleteByScanScroll()` is trivial.
However, `Client::removeDocuments()` is called in UnitOfWork->commit()->commitRemoved(). Passing a refresh flag that far down the call chain feels a bit awkward. Especially since, if the commit() $refresh param is set, it refreshes all the indices already (UnitOfWork lines 161-166).
My uncertainty is about the way `SearchManager::remove()` uses `UnitOfWork::remove()` (where a refresh flag doesn't do anything, unless we pass class/index parameters as well) and `SearchManager::removeAll()` goes directly to `Client::removeAll()` (where the refresh flag can be used easily).
Maybe I'm overthinking this, or misunderstanding your comment. Thoughts?